core/components/range_minimum_query.hpp

```diff
@@ -0,0 +1,708 @@
+                                                      block_size - 1);
+            // from beginning of the block to last
+            const auto last_argmin =
+                last_block_base +
+                block_lut_->lookup(last_block_tree_index, 0, last_local);
+            const auto first_min = values_[first_argmin];
+            const auto last_min = values_[last_argmin];
+            return first_min <= last_min ? query_result{first_argmin, first_min}
+                                         : query_result{last_argmin, last_min};
+        }
+        // general case: both values in different non-adjacent blocks
+        const auto first_full_block =
+            first_local == 0 ? first_block : first_block + 1;
+        const auto last_full_block =
+            last_local == block_size - 1 ? last_block : last_block - 1;
+        const auto full_block_result =
+            superblocks_.query(first_full_block, last_full_block);
+        const auto first_block_argmin =
+            first_block_base + block_lut_->lookup(first_block_tree_index,
+                                                  first_local, block_size - 1);
+        const auto last_block_argmin =
+            last_block_base +
+            block_lut_->lookup(last_block_tree_index, 0, last_local);
+        const auto first_block_min = values_[first_block_argmin];
+        const auto last_block_min = values_[last_block_argmin];
+        auto result_min =
+            min(full_block_result.min, min(first_block_min, last_block_min));
+        const auto get_full_block_result_argmin = [&] {
+            return (full_block_result.argmin * block_size +
+                    block_argmin_.get(full_block_result.argmin));
+        };
+        auto result_argmin = first_block_min == result_min ? first_block_argmin
+                             : full_block_result.min == result_min
+                                 ? get_full_block_result_argmin()
+                                 : last_block_argmin;
+        return query_result{result_argmin, result_min};
+    }
+
+    /** Returns the number of values in the value array. */
+    constexpr index_type size() const { return size_; }
+
+private:
+    index_type num_blocks_;
+    const index_type* values_;
+    const uint16* block_tree_indices_;
+    bit_packed_span<int, index_type, const uint32> block_argmin_;
+    superblock_lookup_type superblocks_;
+    const block_lookup_type* block_lut_;
+    index_type size_;
+};
+
+
+/**
+ * Owning version of range_minimum_query, which creates the necessary data from
+ * an input array automatically.
+ *
+ * @tparam IndexType  the type of indices and values in the underlying array.
+ */
+template <typename IndexType>
+class device_range_minimum_query {
+public:
+    constexpr static int block_size = 8;
+    constexpr static int block_argmin_num_bits =
+        ceil_log2_constexpr(block_size);
+    using index_type = IndexType;
+    using view_type = range_minimum_query<block_size, index_type>;
```

Comment:
nit: I find the naming a bit odd. I would have expected that the `device_*` does not own data, and instead is the view type. But I think that can be consolidated when we go deeper into providing device-side view types for more of our core types.

Reply:
agreed, I think I will swap device_ and non-device names